src/Build/Evaluation/ProjectRootElementCache.cs

```diff
@@ -209,92 +222,115 @@ private bool IsInvalidEntry(string projectFile, ProjectRootElement projectRootEl
         /// If item is found, boosts it to the top of the strong cache.
         /// </remarks>
         /// <param name="projectFile">The project file which contains the ProjectRootElement.  Must be a full path.</param>
-        /// <param name="openProjectRootElement">The delegate to use to load if necessary. May be null. Must not update the cache.</param>
+        /// <param name="loadProjectRootElement">The delegate to use to load if necessary. May be null. Must not update the cache.</param>
         /// <param name="isExplicitlyLoaded"><code>true</code> if the project is explicitly loaded, otherwise <code>false</code>.</param>
         /// <param name="preserveFormatting"><code>true</code> to the project was loaded with the formated preserved, otherwise <code>false</code>.</param>
         /// <returns>The ProjectRootElement instance if one exists.  Null otherwise.</returns>
-        internal override ProjectRootElement Get(string projectFile, OpenProjectRootElement openProjectRootElement, bool isExplicitlyLoaded,
+        internal override ProjectRootElement Get(string projectFile, OpenProjectRootElement loadProjectRootElement, bool isExplicitlyLoaded,
             bool? preserveFormatting)
         {
-            // Should already have been canonicalized
-            ErrorUtilities.VerifyThrowInternalRooted(projectFile);
+#if DEBUG
+            // Verify that loadProjectRootElement delegate does not call ProjectRootElementCache.Get().
+            Interlocked.Increment(ref s_getEntriesNumber);
+            ErrorUtilities.VerifyThrow(
+                s_getEntriesNumber == 1,
```

Comment:
Should you be verifying the return value of `Interlocked.Increment` instead?

Reply:
Good point, there’s a window between the increment and the check where the value could have changed again.

Reply:
Hmm, it seems i was a bit confused when I changed to `Interlocked.Increment`. We do not need it, cause the value `s_getEntriesNumber` is thread static, not just static. Our goal is to catch if we have two `Get()` calls in the call stack (possibly from different instances of `ProjectRootElementCache`). So, there is no chance that value would be changed from some other thread. I will return back to usual increments/decrements here.

Reply:
makes sense